### ISSUE TITLE
Shoule eltype really be defined on ForwardDiffNumbers as it currently is?

### ISSUE THREAD

**Issue by KristofferC** [JuliaDiff/ForwardDiff.jl#94]
I believe the current definition of `eltype` on the subtypes of `ForwardDiffNumber`s is confusing

Numbers in Julia tend to be viewed similar to 1 length containers containing themselves. We can for example do:

```julia
julia> 1[1]
1
```

Putting it that way, it seems that `eltype(Vector{T}) == eltype(T)`, or (equivalently?) `eltype(T) == typeof(T)` should be an identity for any `T <: Number` type. However, for ForwardDiffNumbers, this is not true, c.f (where `G` is a `GradientNumber`):

```jl
julia> eltype(1)
Int64

julia> eltype([1])
Int64

julia> eltype(G)
Float64

julia> eltype([G])
ForwardDiff.GradientNumber{3,Float64,Tuple{Float64,Float64,Float64}}
```

Due to this, code that has been type inferred fine with Base Julia Number types has gotten type inference problems with `ForwardDiffNumbers`.

I propose changing `eltype` to `eltype{T <: ForwardDiffNumber}(::Type{T}) = T` and use some other internal function to get the `T` in `ForwardDiffNumbers{N, T, C}`.

**Comment by KristofferC** [comment]
Also, tell me if I blab too much in too little time :)

**Comment by jrevels** [comment]
> I propose changing eltype to eltype{T <: ForwardDiffNumber}(::Type{T}) = T and use some other internal function to get the T in ForwardDiffNumbers{N, T, C}.

Sounds like a good idea to me. 

> Also, tell me if I blab too much in too little time :)

On the contrary, this is the perfect time for discussion, since I now have some time to spend on refactoring ForwardDiff.

**Comment by jrevels** [comment]
fixed in #102
